In [1]:
#importing basics libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

In [3]:
df=pd.read_excel('flight_price.xlsx')
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [5]:
df.nunique()

Airline              12
Date_of_Journey      44
Source                5
Destination           6
Route               128
Dep_Time            222
Arrival_Time       1343
Duration            368
Total_Stops           5
Additional_Info      10
Price              1870
dtype: int64

In [82]:
df.isna().sum()

Airline               0
Source                0
Destination           0
Duration              0
Total_Stops           0
Additional_Info       0
Price                 0
Date                  0
Month                 0
Year                  0
dep_hour              0
dep_min               0
Arrival_hour          0
Arrival_min           0
duration_hour         0
duration_min       1006
dtype: int64

In [8]:
df.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

### There is one null value in the Total_Stops column

In [9]:
df.shape

(10683, 11)

In [10]:
df.dropna(inplace=True)

In [11]:
df.shape

(10682, 11)

In [14]:
df.duplicated().sum()

220

### there are 220 duplated values in the data set

In [15]:
df.drop_duplicates(inplace=True)

In [17]:
df.shape

(10462, 11)

In [19]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

## Lets clean the columns one by one

In [23]:
df['Date']=df['Date_of_Journey'].str.split('/')

In [24]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  \
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897   
1    05:50         13:15   7h 25m     2 stops         No info   7662   
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882   
3    18:05         23:30   5h 25m      1 stop         No info   6218   
4    16:50         21:35   4h 45m      1 stop         No info  13302   

             Date  
0  [24, 03, 2019]  
1   [1, 05, 2019]  
2   [9, 06, 2019]  
3  [12, 05, 2019]  
4  [01, 03, 2019]

In [27]:
df['Date']=df['Date_of_Journey'].str.split('/').str[0]
df['Month']=df['Date_of_Journey'].str.split('/').str[1]
df['Year']=df['Date_of_Journey'].str.split('/').str[2]

In [28]:
df.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price Date  \
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897   24   
1    05:50         13:15   7h 25m     2 stops         No info   7662    1   
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882    9   
3    18:05         23:30   5h 25m      1 stop         No info   6218   12   
4    16:50         21:35   4h 45m      1 stop         No info  13302   01   

  Month  Year  
0    03  2019  
1    05  2019  
2    06  2019  
3    05  2019  
4    03  2019

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Route            10462 non-null  object
 4   Dep_Time         10462 non-null  object
 5   Arrival_Time     10462 non-null  object
 6   Duration         10462 non-null  object
 7   Total_Stops      10462 non-null  object
 8   Additional_Info  10462 non-null  object
 9   Price            10462 non-null  int64 
 10  Date             10462 non-null  object
 11  Month            10462 non-null  object
 12  Year             10462 non-null  object
dtypes: int64(1), object(12)
memory usage: 1.4+ MB


In [34]:
df['Date']=df['Date'].astype(int)
df['Month']=df['Month'].astype(int)
df['Year']=df['Year'].astype(int)


In [35]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Route            10462 non-null  object
 4   Dep_Time         10462 non-null  object
 5   Arrival_Time     10462 non-null  object
 6   Duration         10462 non-null  object
 7   Total_Stops      10462 non-null  object
 8   Additional_Info  10462 non-null  object
 9   Price            10462 non-null  int64 
 10  Date             10462 non-null  int32 
 11  Month            10462 non-null  int32 
 12  Year             10462 non-null  int32 
dtypes: int32(3), int64(1), object(9)
memory usage: 1.2+ MB


In [30]:
df.drop('Date_of_Journey',inplace=True,axis=1)

In [31]:
df.head()

Airline    Source Destination                  Route Dep_Time  \
0       IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL    16:50   

   Arrival_Time Duration Total_Stops Additional_Info  Price Date Month  Year  
0  01:10 22 Mar   2h 50m    non-stop         No info   3897   24    03  2019  
1         13:15   7h 25m     2 stops         No info   7662    1    05  2019  
2  04:25 10 Jun      19h     2 stops         No info  13882    9    06  2019  
3         23:30   5h 25m      1 stop         No info   6218   12    05  2019  
4         21:35   4h 45m      1 stop         No info  13302   01    03  2019

In [36]:
df['Destination'].unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [37]:
df.drop('Route',axis=1,inplace=True)

In [38]:
df.head()

Airline    Source Destination Dep_Time  Arrival_Time Duration  \
0       IndiGo  Banglore   New Delhi    22:20  01:10 22 Mar   2h 50m   
1    Air India   Kolkata    Banglore    05:50         13:15   7h 25m   
2  Jet Airways     Delhi      Cochin    09:25  04:25 10 Jun      19h   
3       IndiGo   Kolkata    Banglore    18:05         23:30   5h 25m   
4       IndiGo  Banglore   New Delhi    16:50         21:35   4h 45m   

  Total_Stops Additional_Info  Price  Date  Month  Year  
0    non-stop         No info   3897    24      3  2019  
1     2 stops         No info   7662     1      5  2019  
2     2 stops         No info  13882     9      6  2019  
3      1 stop         No info   6218    12      5  2019  
4      1 stop         No info  13302     1      3  2019

## cleaning the Dep_Time column

In [39]:
df['dep_hour']=df['Dep_Time'].str.split(':').str[0]
df['dep_min']=df['Dep_Time'].str.split(':').str[1]

In [40]:
df.head()

Airline    Source Destination Dep_Time  Arrival_Time Duration  \
0       IndiGo  Banglore   New Delhi    22:20  01:10 22 Mar   2h 50m   
1    Air India   Kolkata    Banglore    05:50         13:15   7h 25m   
2  Jet Airways     Delhi      Cochin    09:25  04:25 10 Jun      19h   
3       IndiGo   Kolkata    Banglore    18:05         23:30   5h 25m   
4       IndiGo  Banglore   New Delhi    16:50         21:35   4h 45m   

  Total_Stops Additional_Info  Price  Date  Month  Year dep_hour dep_min  
0    non-stop         No info   3897    24      3  2019       22      20  
1     2 stops         No info   7662     1      5  2019       05      50  
2     2 stops         No info  13882     9      6  2019       09      25  
3      1 stop         No info   6218    12      5  2019       18      05  
4      1 stop         No info  13302     1      3  2019       16      50

In [41]:
df.drop('Dep_Time',inplace=True,axis=1)

In [42]:
df.head()

Airline    Source Destination  Arrival_Time Duration Total_Stops  \
0       IndiGo  Banglore   New Delhi  01:10 22 Mar   2h 50m    non-stop   
1    Air India   Kolkata    Banglore         13:15   7h 25m     2 stops   
2  Jet Airways     Delhi      Cochin  04:25 10 Jun      19h     2 stops   
3       IndiGo   Kolkata    Banglore         23:30   5h 25m      1 stop   
4       IndiGo  Banglore   New Delhi         21:35   4h 45m      1 stop   

  Additional_Info  Price  Date  Month  Year dep_hour dep_min  
0         No info   3897    24      3  2019       22      20  
1         No info   7662     1      5  2019       05      50  
2         No info  13882     9      6  2019       09      25  
3         No info   6218    12      5  2019       18      05  
4         No info  13302     1      3  2019       16      50

In [43]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Arrival_Time     10462 non-null  object
 4   Duration         10462 non-null  object
 5   Total_Stops      10462 non-null  object
 6   Additional_Info  10462 non-null  object
 7   Price            10462 non-null  int64 
 8   Date             10462 non-null  int32 
 9   Month            10462 non-null  int32 
 10  Year             10462 non-null  int32 
 11  dep_hour         10462 non-null  object
 12  dep_min          10462 non-null  object
dtypes: int32(3), int64(1), object(9)
memory usage: 1.2+ MB


In [44]:
df['dep_hour']=df['dep_hour'].astype(int)
df['dep_min']=df['dep_min'].astype(int)

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Arrival_Time     10462 non-null  object
 4   Duration         10462 non-null  object
 5   Total_Stops      10462 non-null  object
 6   Additional_Info  10462 non-null  object
 7   Price            10462 non-null  int64 
 8   Date             10462 non-null  int32 
 9   Month            10462 non-null  int32 
 10  Year             10462 non-null  int32 
 11  dep_hour         10462 non-null  int32 
 12  dep_min          10462 non-null  int32 
dtypes: int32(5), int64(1), object(7)
memory usage: 1.2+ MB


## cleaning the Arrival Time column

In [46]:
df.head()

Airline    Source Destination  Arrival_Time Duration Total_Stops  \
0       IndiGo  Banglore   New Delhi  01:10 22 Mar   2h 50m    non-stop   
1    Air India   Kolkata    Banglore         13:15   7h 25m     2 stops   
2  Jet Airways     Delhi      Cochin  04:25 10 Jun      19h     2 stops   
3       IndiGo   Kolkata    Banglore         23:30   5h 25m      1 stop   
4       IndiGo  Banglore   New Delhi         21:35   4h 45m      1 stop   

  Additional_Info  Price  Date  Month  Year  dep_hour  dep_min  
0         No info   3897    24      3  2019        22       20  
1         No info   7662     1      5  2019         5       50  
2         No info  13882     9      6  2019         9       25  
3         No info   6218    12      5  2019        18        5  
4         No info  13302     1      3  2019        16       50

In [70]:
df['Arrival_hour']=df['Arrival_Time'].str.split(':').str[0]
df['arrival_min']=df['Arrival_Time'].str.split(':').str[1]
df['Arrival_min']=df['arrival_min'].str.split(' ').str[0]
df.drop('Arrival_Time',inplace=True,axis=1)



In [71]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min Arrival_hour Arrival_min  \
0   3897    24      3  2019        22       20           01          10   
1   7662     1      5  2019         5       50           13          15   
2  13882     9      6  2019         9       25           04          25   
3   6218    12      5  2019        18        5           23          30   
4  13302     1      3  2019        16       50           21          35   

  arrival_min  
0   10 22 Mar  
1          15  
2   25 10 Jun  
3          30  
4          35

In [72]:
df.drop('arrival_min',inplace=True,axis=1)

In [73]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min Arrival_hour Arrival_min  
0   3897    24      3  2019        22       20           01          10  
1   7662     1      5  2019         5       50           13          15  
2  13882     9      6  2019         9       25           04          25  
3   6218    12      5  2019        18        5           23          30  
4  13302     1      3  2019        16       50           21          35

In [74]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Duration         10462 non-null  object
 4   Total_Stops      10462 non-null  object
 5   Additional_Info  10462 non-null  object
 6   Price            10462 non-null  int64 
 7   Date             10462 non-null  int32 
 8   Month            10462 non-null  int32 
 9   Year             10462 non-null  int32 
 10  dep_hour         10462 non-null  int32 
 11  dep_min          10462 non-null  int32 
 12  Arrival_hour     10462 non-null  object
 13  Arrival_min      10462 non-null  object
dtypes: int32(5), int64(1), object(8)
memory usage: 1.2+ MB


In [76]:
df['Arrival_hour']=df['Arrival_hour'].astype(int)
df['Arrival_min']=df['Arrival_min'].astype(int)

In [77]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Duration         10462 non-null  object
 4   Total_Stops      10462 non-null  object
 5   Additional_Info  10462 non-null  object
 6   Price            10462 non-null  int64 
 7   Date             10462 non-null  int32 
 8   Month            10462 non-null  int32 
 9   Year             10462 non-null  int32 
 10  dep_hour         10462 non-null  int32 
 11  dep_min          10462 non-null  int32 
 12  Arrival_hour     10462 non-null  int32 
 13  Arrival_min      10462 non-null  int32 
dtypes: int32(7), int64(1), object(6)
memory usage: 1.2+ MB


In [78]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  
0   3897    24      3  2019        22       20             1           10  
1   7662     1      5  2019         5       50            13           15  
2  13882     9      6  2019         9       25             4           25  
3   6218    12      5  2019        18        5            23           30  
4  13302     1      3  2019        16       50            21           35

## Cleaning the Duration column

In [80]:
df['duration_hour']=df['Duration'].str.split(' ').str[0]
df['duration_min']=df['Duration'].str.split(' ').str[1]

In [81]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  \
0   3897    24      3  2019        22       20             1           10   
1   7662     1      5  2019         5       50            13           15   
2  13882     9      6  2019         9       25             4           25   
3   6218    12      5  2019        18        5            23           30   
4  13302     1      3  2019        16       50            21           35   

  duration_hour duration_min  
0            2h          50m  
1            7h          25m  
2           19h          NaN  
3            5h          25m  
4            4h          45m

In [85]:
df['Duration_min']=df['duration_min'].fillna('0m')

In [86]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  \
0   3897    24      3  2019        22       20             1           10   
1   7662     1      5  2019         5       50            13           15   
2  13882     9      6  2019         9       25             4           25   
3   6218    12      5  2019        18        5            23           30   
4  13302     1      3  2019        16       50            21           35   

  duration_hour duration_min Duration_min  
0            2h          50m          50m  
1            7h          25m          25m  
2           19h          NaN           0m  
3            5h          25m          25m  
4            4h          45m          45m

In [87]:
df.drop('duration_min',inplace=True,axis=1)

In [88]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  \
0   3897    24      3  2019        22       20             1           10   
1   7662     1      5  2019         5       50            13           15   
2  13882     9      6  2019         9       25             4           25   
3   6218    12      5  2019        18        5            23           30   
4  13302     1      3  2019        16       50            21           35   

  duration_hour Duration_min  
0            2h          50m  
1            7h          25m  
2           19h           0m  
3            5h          25m  
4            4h          45m

In [89]:
df['Duration_min'].unique()

array(['50m', '25m', '0m', '45m', '30m', '5m', '15m', '35m', '10m', '20m',
       '55m', '40m'], dtype=object)

In [94]:
df['duration_hr']=df['duration_hour'].str.split('h').str[0]
df['duration_min']=df['Duration_min'].str.split('m').str[0]

In [95]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  \
0   3897    24      3  2019        22       20             1           10   
1   7662     1      5  2019         5       50            13           15   
2  13882     9      6  2019         9       25             4           25   
3   6218    12      5  2019        18        5            23           30   
4  13302     1      3  2019        16       50            21           35   

  duration_hour Duration_min duration_hr duration_min  
0            2h          50m           2           50  
1            7h          25m           7           25  
2           19h           0m          19            0  
3            5h          25m           5           25  
4            4h          45m           4           45

In [97]:
df['duration_hr'].unique()


array(['2', '7', '19', '5', '4', '15', '21', '25', '13', '12', '26', '22',
       '23', '20', '10', '6', '11', '8', '16', '3', '27', '1', '14', '9',
       '18', '17', '24', '30', '28', '29', '37', '34', '38', '35', '36',
       '47', '33', '32', '31', '42', '39', '5m', '41', '40'], dtype=object)

In [98]:
df['duration_min'].unique()

array(['50', '25', '0', '45', '30', '5', '15', '35', '10', '20', '55',
       '40'], dtype=object)

In [99]:
df.drop('duration_hour',inplace=True,axis=1)
df.drop('Duration_min',inplace=True,axis=1)

In [100]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  \
0   3897    24      3  2019        22       20             1           10   
1   7662     1      5  2019         5       50            13           15   
2  13882     9      6  2019         9       25             4           25   
3   6218    12      5  2019        18        5            23           30   
4  13302     1      3  2019        16       50            21           35   

  duration_hr duration_min  
0           2           50  
1           7           25  
2          19            0  
3           5           25  
4           4           45

In [101]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Duration         10462 non-null  object
 4   Total_Stops      10462 non-null  object
 5   Additional_Info  10462 non-null  object
 6   Price            10462 non-null  int64 
 7   Date             10462 non-null  int32 
 8   Month            10462 non-null  int32 
 9   Year             10462 non-null  int32 
 10  dep_hour         10462 non-null  int32 
 11  dep_min          10462 non-null  int32 
 12  Arrival_hour     10462 non-null  int32 
 13  Arrival_min      10462 non-null  int32 
 14  duration_hr      10462 non-null  object
 15  duration_min     10462 non-null  object
dtypes: int32(7), int64(1), object(8)
memory usage: 1.3+ MB


In [103]:
df['duration_min']=df['duration_min'].astype(int)
# df['duration_hr']=df['duration_hr'].astype(int)

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10462 entries, 0 to 10682
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10462 non-null  object
 1   Source           10462 non-null  object
 2   Destination      10462 non-null  object
 3   Duration         10462 non-null  object
 4   Total_Stops      10462 non-null  object
 5   Additional_Info  10462 non-null  object
 6   Price            10462 non-null  int64 
 7   Date             10462 non-null  int32 
 8   Month            10462 non-null  int32 
 9   Year             10462 non-null  int32 
 10  dep_hour         10462 non-null  int32 
 11  dep_min          10462 non-null  int32 
 12  Arrival_hour     10462 non-null  int32 
 13  Arrival_min      10462 non-null  int32 
 14  duration_hr      10462 non-null  object
 15  duration_min     10462 non-null  int32 
dtypes: int32(8), int64(1), object(7)
memory usage: 1.3+ MB


In [106]:
df.head()

Airline    Source Destination Duration Total_Stops Additional_Info  \
0       IndiGo  Banglore   New Delhi   2h 50m    non-stop         No info   
1    Air India   Kolkata    Banglore   7h 25m     2 stops         No info   
2  Jet Airways     Delhi      Cochin      19h     2 stops         No info   
3       IndiGo   Kolkata    Banglore   5h 25m      1 stop         No info   
4       IndiGo  Banglore   New Delhi   4h 45m      1 stop         No info   

   Price  Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  \
0   3897    24      3  2019        22       20             1           10   
1   7662     1      5  2019         5       50            13           15   
2  13882     9      6  2019         9       25             4           25   
3   6218    12      5  2019        18        5            23           30   
4  13302     1      3  2019        16       50            21           35   

  duration_hr  duration_min  
0           2            50  
1           7            25  
2          19             0  
3           5            25  
4           4            45

In [107]:
df.drop('Duration',axis=1,inplace=True)

In [108]:
df.head()

Airline    Source Destination Total_Stops Additional_Info  Price  Date  \
0       IndiGo  Banglore   New Delhi    non-stop         No info   3897    24   
1    Air India   Kolkata    Banglore     2 stops         No info   7662     1   
2  Jet Airways     Delhi      Cochin     2 stops         No info  13882     9   
3       IndiGo   Kolkata    Banglore      1 stop         No info   6218    12   
4       IndiGo  Banglore   New Delhi      1 stop         No info  13302     1   

   Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min duration_hr  \
0      3  2019        22       20             1           10           2   
1      5  2019         5       50            13           15           7   
2      6  2019         9       25             4           25          19   
3      5  2019        18        5            23           30           5   
4      3  2019        16       50            21           35           4   

   duration_min  
0            50  
1            25  
2             0  
3            25  
4            45

## cleaning the Total_Stops column

In [109]:
df['Total_Stops'].unique()

array(['non-stop', '2 stops', '1 stop', '3 stops', '4 stops'],
      dtype=object)

### map function is used for data encoding

In [110]:
df['Total_Stops']=df['Total_Stops'].map({'non-stop':0, '2 stops':2, '1 stop':1, '3 stops':3, '4 stops':4})

In [111]:
df['Total_Stops'].unique()

array([0, 2, 1, 3, 4], dtype=int64)

In [112]:
df.head()

Airline    Source Destination  Total_Stops Additional_Info  Price  \
0       IndiGo  Banglore   New Delhi            0         No info   3897   
1    Air India   Kolkata    Banglore            2         No info   7662   
2  Jet Airways     Delhi      Cochin            2         No info  13882   
3       IndiGo   Kolkata    Banglore            1         No info   6218   
4       IndiGo  Banglore   New Delhi            1         No info  13302   

   Date  Month  Year  dep_hour  dep_min  Arrival_hour  Arrival_min  \
0    24      3  2019        22       20             1           10   
1     1      5  2019         5       50            13           15   
2     9      6  2019         9       25             4           25   
3    12      5  2019        18        5            23           30   
4     1      3  2019        16       50            21           35   

  duration_hr  duration_min  
0           2            50  
1           7            25  
2          19             0  
3           5            25  
4           4            45

In [113]:
df.drop('Additional_Info',inplace=True,axis=1)

In [114]:
df.head()

Airline    Source Destination  Total_Stops  Price  Date  Month  Year  \
0       IndiGo  Banglore   New Delhi            0   3897    24      3  2019   
1    Air India   Kolkata    Banglore            2   7662     1      5  2019   
2  Jet Airways     Delhi      Cochin            2  13882     9      6  2019   
3       IndiGo   Kolkata    Banglore            1   6218    12      5  2019   
4       IndiGo  Banglore   New Delhi            1  13302     1      3  2019   

   dep_hour  dep_min  Arrival_hour  Arrival_min duration_hr  duration_min  
0        22       20             1           10           2            50  
1         5       50            13           15           7            25  
2         9       25             4           25          19             0  
3        18        5            23           30           5            25  
4        16       50            21           35           4            45

In [117]:
df['duration_hr'].unique()

array(['2', '7', '19', '5', '4', '15', '21', '25', '13', '12', '26', '22',
       '23', '20', '10', '6', '11', '8', '16', '3', '27', '1', '14', '9',
       '18', '17', '24', '30', '28', '29', '37', '34', '38', '35', '36',
       '47', '33', '32', '31', '42', '39', '5m', '41', '40'], dtype=object)

In [119]:
# Drop rows where 'values' is '5m'
df = df[df['duration_hr'] != '5m']


In [120]:
df.reset_index(drop=True, inplace=True)


In [121]:
df['duration_hr'].unique()

array(['2', '7', '19', '5', '4', '15', '21', '25', '13', '12', '26', '22',
       '23', '20', '10', '6', '11', '8', '16', '3', '27', '1', '14', '9',
       '18', '17', '24', '30', '28', '29', '37', '34', '38', '35', '36',
       '47', '33', '32', '31', '42', '39', '41', '40'], dtype=object)

In [122]:
df['duration_hr']=df['duration_hr'].astype(int)


In [123]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10461 entries, 0 to 10460
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Airline       10461 non-null  object
 1   Source        10461 non-null  object
 2   Destination   10461 non-null  object
 3   Total_Stops   10461 non-null  int64 
 4   Price         10461 non-null  int64 
 5   Date          10461 non-null  int32 
 6   Month         10461 non-null  int32 
 7   Year          10461 non-null  int32 
 8   dep_hour      10461 non-null  int32 
 9   dep_min       10461 non-null  int32 
 10  Arrival_hour  10461 non-null  int32 
 11  Arrival_min   10461 non-null  int32 
 12  duration_hr   10461 non-null  int32 
 13  duration_min  10461 non-null  int32 
dtypes: int32(9), int64(2), object(3)
memory usage: 776.5+ KB


In [125]:
df['Airline'].unique()

array(['IndiGo', 'Air India', 'Jet Airways', 'SpiceJet',
       'Multiple carriers', 'GoAir', 'Vistara', 'Air Asia',
       'Vistara Premium economy', 'Jet Airways Business',
       'Multiple carriers Premium economy', 'Trujet'], dtype=object)

In [126]:
from sklearn.preprocessing import OneHotEncoder

In [127]:
encoder=OneHotEncoder()

In [128]:
encoder.fit_transform(df[['Airline','Source','Destination']]).toarray()

array([[0., 0., 0., ..., 0., 0., 1.],
       [0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 1., 0., ..., 0., 0., 0.]])

In [131]:
df1=pd.DataFrame(encoder.fit_transform(df[['Airline','Source','Destination']]).toarray(),columns=encoder.get_feature_names_out())

In [132]:
df1

Airline_Air Asia  Airline_Air India  Airline_GoAir  Airline_IndiGo  \
0                   0.0                0.0            0.0             1.0   
1                   0.0                1.0            0.0             0.0   
2                   0.0                0.0            0.0             0.0   
3                   0.0                0.0            0.0             1.0   
4                   0.0                0.0            0.0             1.0   
...                 ...                ...            ...             ...   
10456               1.0                0.0            0.0             0.0   
10457               0.0                1.0            0.0             0.0   
10458               0.0                0.0            0.0             0.0   
10459               0.0                0.0            0.0             0.0   
10460               0.0                1.0            0.0             0.0   

       Airline_Jet Airways  Airline_Jet Airways Business  \
0                      0.0                           0.0   
1                      0.0                           0.0   
2                      1.0                           0.0   
3                      0.0                           0.0   
4                      0.0                           0.0   
...                    ...                           ...   
10456                  0.0                           0.0   
10457                  0.0                           0.0   
10458                  1.0                           0.0   
10459                  0.0                           0.0   
10460                  0.0                           0.0   

       Airline_Multiple carriers  Airline_Multiple carriers Premium economy  \
0                            0.0                                        0.0   
1                            0.0                                        0.0   
2                            0.0                                        0.0   
3                            0.0                                        0.0   
4                            0.0                                        0.0   
...                          ...                                        ...   
10456                        0.0                                        0.0   
10457                        0.0                                        0.0   
10458                        0.0                                        0.0   
10459                        0.0                                        0.0   
10460                        0.0                                        0.0   

       Airline_SpiceJet  Airline_Trujet  ...  Source_Chennai  Source_Delhi  \
0                   0.0             0.0  ...             0.0           0.0   
1                   0.0             0.0  ...             0.0           0.0   
2                   0.0             0.0  ...             0.0           1.0   
3                   0.0             0.0  ...             0.0           0.0   
4                   0.0             0.0  ...             0.0           0.0   
...                 ...             ...  ...             ...           ...   
10456               0.0             0.0  ...             0.0           0.0   
10457               0.0             0.0  ...             0.0           0.0   
10458               0.0             0.0  ...             0.0           0.0   
10459               0.0             0.0  ...             0.0           0.0   
10460               0.0             0.0  ...             0.0           1.0   

       Source_Kolkata  Source_Mumbai  Destination_Banglore  \
0                 0.0            0.0                   0.0   
1                 1.0            0.0                   1.0   
2                 0.0            0.0                   0.0   
3                 1.0            0.0                   1.0   
4                 0.0            0.0                   0.0   
...               ...            ...                   ...   
10456             1.0            0.0     

In [136]:
df.head()

Airline    Source Destination  Total_Stops  Price  Date  Month  Year  \
0       IndiGo  Banglore   New Delhi            0   3897    24      3  2019   
1    Air India   Kolkata    Banglore            2   7662     1      5  2019   
2  Jet Airways     Delhi      Cochin            2  13882     9      6  2019   
3       IndiGo   Kolkata    Banglore            1   6218    12      5  2019   
4       IndiGo  Banglore   New Delhi            1  13302     1      3  2019   

   dep_hour  dep_min  Arrival_hour  Arrival_min  duration_hr  duration_min  
0        22       20             1           10            2            50  
1         5       50            13           15            7            25  
2         9       25             4           25           19             0  
3        18        5            23           30            5            25  
4        16       50            21           35            4            45

In [146]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10461 entries, 0 to 10460
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Airline       10461 non-null  object
 1   Source        10461 non-null  object
 2   Destination   10461 non-null  object
 3   Total_Stops   10461 non-null  int64 
 4   Price         10461 non-null  int64 
 5   Date          10461 non-null  int32 
 6   Month         10461 non-null  int32 
 7   Year          10461 non-null  int32 
 8   dep_hour      10461 non-null  int32 
 9   dep_min       10461 non-null  int32 
 10  Arrival_hour  10461 non-null  int32 
 11  Arrival_min   10461 non-null  int32 
 12  duration_hr   10461 non-null  int32 
 13  duration_min  10461 non-null  int32 
dtypes: int32(9), int64(2), object(3)
memory usage: 776.5+ KB
